Enable multi-row INSERTs

jasonmp85 opened the pull request:
Fixes #455

This is a pretty substantial refactoring of the existing modify path within the router executor and planner. In particular, we now hunt for all VALUES range table entries in `INSERT` statements and group the rows contained therein by shard identifier. These rows are stashed away for later in "ModifyRoute" elements. During deparse, the appropriate RTE is extracted from the Query and its values list is replaced by these rows before any SQL is generated.

In this way, we can create multiple Tasks, but only one per shard, to piecemeal execute a multi-row `INSERT`. The execution of jobs containing such tasks now exclusively go through the "multi-router executor" which was previously used for e.g. `INSERT INTO ... SELECT`.

By piggybacking onto that executor, we participate in ongoing transactions, get rollback-ability, etc. In short order, the only remaining use of the "single modify" router executor will be for bare single-row INSERT statements (i.e. those not in a transaction).

This change appropriately handles deferred pruning as well as master-evaluated functions.

metdos commented:
Just noting that I'm also working on modify path especially on UPDATE/DELETE, and there will be some conflicts with this PR #1483. It seems that most likely I will need to rebase :)

marcocitus commented on src/backend/distributed/executor/multi_executor.c:
```diff
@@ -66,7 +66,8 @@ static CustomExecMethods RouterSingleModifyCustomExecMethods = {
 
-static CustomExecMethods RouterMultiModifyCustomExecMethods = {
+/* not static to enable swapping in multi-modify logic during router execution */
+CustomExecMethods RouterMultiModifyCustomExecMethods = {
```
It seems the only distinction between `RouterSingleModifyCustomExecMethods` and `RouterMultiModifyCustomExecMethods` is in the `ExecCustomScan` and the only metric used to choose between them is the number of tasks. What if we collapse them into one `ExecMethods` with a `RouterModifyExecScan` that calls the right function based on the task count?

jasonmp85 replied:
Hm. I'm not sure that's correct, at least considering deadlock detection-adjacent features. I was led to believe that those changes were going to permit arbitrary multi-shard/multi-statement transactions. In that case, we want to (a) use the multi-task executor (and its transaction machinery) _even_ in cases where the modification hits only a single shard, or (b) use the single-shard path for single-shard modifications outside of a transaction block.

Does that make sense? I'm hesitant to make the suggested change until the other case is cleared up.

jasonmp85 replied:
Leaving this as-is, at least for the time being.

marcocitus replied:
We're not doing anything about multi-shard update/delete for the moment, but I guess there could be a future in which multiple functions differ. Fine leaving this as is.

anarazel replied:
This seems like a bad idea. What if there's other queries in-progress that refer to `RouterMultiModifyCustomExecMethods`? With cursors there could be many custom scan nodes referencing `RouterMultiModifyCustomExecMethods`.

anarazel replied:
Err, misunderstood what you meant with "swapping in multi-modify logic" - thought you were replacing an individual callback in it. Sorry for the noise.  Why not just say 'so it can be referenced by the multi-modify' logic?

jasonmp85 replied:
I meant it can be swapped in, i.e. the compound verb "swap in". I realized later there was a garden path parse of this phrasing, but held off modifying it until someone complained. Looks like I'll change it now!

marcocitus commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -429,10 +429,19 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
 		PlanState *planState = &(scanState->customScanState.ss.ps);
 
 		ExecuteMasterEvaluableFunctions(jobQuery, planState);
+		ExecuteTaskListFunctions(taskList, planState);
```
Shouldn't it be sufficient to evaluate functions in the `jobQuery` and then break down the jobQuery into multiple tasks?

jasonmp85 replied:
I guess? But isn't that, uh, planning? The lines between planning and execution are getting real blurry here. The planning phase (when possible) prunes and groups all rows by their destination shards. Those groups become tasks, whose regrouped rows we use to generate the final SQL.

What you suggest I could see, but it would drive a wedge between what we do in planning and what we do in execution (i.e. if fully plannable ahead of time, we'd create these tasks like I do now, but if not, we'd build things directly from the Query at execution).

marcocitus replied:
I think we should generate tasks when we're ready to generate the correct number of tasks, which might only be at execution time (e.g. function in the partition column). I guess there is an important distinction between queries with a function call in the partition column, and queries with a constant in the partition column, but a function call in another column. In the former case we cannot generate the correct number of tasks at planning time (use `ExecuteMasterEvaluableFunctions(jobQuery, planState)`?), while in the latter case we can (use `ExecuteTaskListFunctions(taskList, planState)`?), though I don't know if we get enough performance benefit out of building the task list upfront to justify the additional function.

In any case we shouldn't do function evaluation twice, that would give unexpected results (e.g. `nextval` getting called twice).

marcocitus commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1015,73 +1005,116 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
-		/* set the flag */
-		modifyTask->upsertQuery = true;
+			/* set the flag */
+			modifyTask->upsertQuery = true;
```
not this PR's fault, but this comment doesn't seem very helpful

jasonmp85 replied:
This comment is :trophy:-winning in its redundancy is what it is.

marcocitus commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1015,73 +1005,116 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
+
+			modifyTask->anchorShardId = shardId;
+			modifyTask->valuesRTEIndex = modifyRoute->partitionvarno;
+			modifyTask->rowValuesLists = modifyRoute->rowValuesLists;
```
Can we avoid storing the `partitionvarno` and `rowValuesList` in `Task` by passing it to `UpdateTaskQueryString`? After we do the deparsing in `UpdateTaskQueryString` we don't need the routes anymore. When we defer pruning to the executor we rebuild the list of routes anyway.

I guess the case in which storing them might be beneficial is when the partition column values are known, but some other columns contain function calls. In that case we can build the list of tasks, but we cannot deparse them yet. On the other hand, we're deparsing them on the line below, so we'll actually do deparsing twice, which is really expensive. It seems simpler and equally performant to just build the tasks in the executor after function evaluation in case there are function calls.

jasonmp85 replied:
I think we could, but what would you propose be done about `RebuildShardQueries` or whatever in the executor? Right here we have both values at hand in a loop, but there we just have access to a possibly-updated list of `Task`s. Having the tasks each know about how to find and replace the rows it needs interpolated in the original query works well in that function. Thoughts?

marcocitus replied:
Fair point, I guess we could *always* rebuild the task list and then do deparsing on-the-fly, but for creating (and preparing) plans with a partially populated task list we need these fields.

marcocitus commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1384,35 +1551,74 @@ ExtractFirstDistributedTableId(Query *query)
 	}
 
-	return targetEntry->expr;
+	/* We've got a multi-row INSERT. Extract all VALUES */
```
Might help elaborating a bit on how multi-row INSERTs are represented in postgres.

jasonmp85 replied:
Sure.

marcocitus commented on src/backend/distributed/utils/citus_clauses.c:
```diff
@@ -141,12 +146,39 @@ ExecuteMasterEvaluableFunctions(Query *query, PlanState *planState)
+		{
+			ListCell *exprListCell = NULL;
+
+			foreach(exprListCell, rte->values_lists)
```
Would be nice to move this into its own function.

jasonmp85 replied:
Sure, sounds good.

marcocitus commented on src/backend/distributed/utils/citus_clauses.c:
```diff
@@ -141,12 +146,39 @@ ExecuteMasterEvaluableFunctions(Query *query, PlanState *planState)
+
+					if (commandType == CMD_INSERT && !insertSelectQuery)
+					{
+						modifiedNode = EvaluateNodeIfReferencesFunction((Node *) expr,
```
Small note: we'll probably stop special casing INSERTs here (#1506). It seems to be a performance hack, which actually causes some bugs / limitations.

jasonmp85 replied:
Ack; fixed here, too.

marcocitus commented on src/test/regress/sql/multi_modifications.sql:
```diff
@@ -146,8 +146,10 @@ DELETE FROM limit_orders WHERE id = 246 AND bidder_id = (random() * 1000);
+-- commands with multiple rows are supported
+INSERT INTO limit_orders VALUES (2037, 'GOOG', 5634, now(), 'buy', random()),
+                                (2038, 'GOOG', 5634, now(), 'buy', random()),
+                                (2039, 'GOOG', 5634, now(), 'buy', random());
```
would be good to add some regression tests in various types of transactions, prepared statements, and upserts.

jasonmp85 replied:
Agreed, was working on them last week but haven't gotten them to a pushable state yet. Will do today (SF time).

marcocitus commented on src/include/distributed/multi_router_planner.h:
```diff
@@ -21,6 +21,14 @@
 #include "nodes/parsenodes.h"
 
 
+typedef struct ModifyRoute
```
Could use a comment on what we use this for.

jasonmp85 replied:
Sure.

marcocitus commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1015,73 +1005,116 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
+			}
+		}
+
+		if (modifyRoute != NULL)
```
Having a partially populated task that we complete in the executor was a bit of the hack to get function evaluation in the partition column. It probably makes more sense to leave the route / task list empty when shard pruning is not yet possible and rebuild the task list in the executor.

e.g.
https://github.com/citusdata/citus/pull/1474/files#diff-4c4f0976727d06684376ecb22e8f5362R1123
https://github.com/citusdata/citus/pull/1474/files#diff-94b1a72e077cfc917cafbafd2a8b80ccR409

jasonmp85 commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1104,154 +1137,280 @@ TargetShardIntervalForModify(Oid distributedTableId, Query *query,
-static ShardInterval *
-FindShardForUpdateOrDelete(Query *query, DeferredErrorMessage **planningError)
+static ModifyRoute *
+BuildRouteForUpdateOrDelete(Query *query, DeferredErrorMessage **planningError)
```
Yeah it was mostly a consistency thing. I didn't want the two paths (`INSERT` vs `UPDATE`/`DELETE`) to have different return types. It only returns one route right now but it makes the rest of the code flow better. Is there anything you wanted me to do here in the code?

jasonmp85 replied:
Leaving as-is.

jasonmp85 commented:
Hey, so @marcocitus I was attempting to deal with #1474 by rebasing it on top of my changes, but some of your comments here sound as though you'd rather me rebase _this_ PR on top of #1474. Which would make more sense?

I think each independently trying to do some of the things you're describing and independently merging to `master` will be needlessly confusing/complicated.

jasonmp85 commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -429,10 +429,19 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
 		if (deferredPruning)
 		{
-			AssignInsertTaskShardId(jobQuery, taskList);
+			AssignInsertJobShardIds(workerJob);
```
Thinking I'm going to need to either merge #1474 to `master` first or rebase _this_ on top of _that_ to satisfy these suggestions.

jasonmp85 commented:
(Alternatively, if #1474 is quickly merged to master I can just rebase atop that rather than keep both PRs open).

jasonmp85 commented:
Rebased and got this to compile, but it's a bit of mess logic-wise (there were > 100 conflicts in the router). Pushing it up here even though it definitely is not in a usable state; I just want a nice GitHub view of the changes for a sec, then I'll be doing some modifications to property apply your suggestions in re: using new functions/flow from #1474…

jasonmp85 commented:
OK @marcocitus, I've made the fewest changes atop this as possible to use the things you added in #1474. Because of the separation of concerns throughout various functions, it made sense to at minimum keep the values lists within the `Task`. Besides, I think there is something to be said for that case where non-partition columns need master evaluation so pruning can be accomplished just not deparsing until last-minute.

I moved the `valuesRTEIndex` into `Job` as it's a single value for any given multi-row query.

If you can give this one more look and organization/style/flow feedback, I'll put the final touches on it tomorrow (and flesh out the tests as I've been promising for some time), and maybe we can get this to merge by Wednesday morning (Pacific). Thanks!

marcocitus commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1925,6 +1786,356 @@ IntersectPlacementList(List *lhsPlacementList, List *rhsPlacementList)
+		RangeTblEntry *referencedRTE = NULL;
+		ListCell *valuesListCell = NULL;
+
+		*partitionvarno = partitionVar->varno;
```
This seems pretty cheap to compute from a Query *. Instead of storing it in the `Job` and passing it as a parameter to several functions, could we call something like `PartitionColumnTargetListIndex(query)` when needed?

jasonmp85 replied:
I'm so glad you agree because I hated adding it to all the functions. Removing.

marcocitus commented:
Code looks good. With the additional regression tests this is a :ship:.

One thing to note is that multi-shard INSERTs will  use the aggressive shard resource locks that we currently rely on to prevent deadlocks between multi-shard commands, which means we only allow one multi-row INSERT per shard at a time, i.e. multi-row INSERTs only run concurrently if they don't modify the same shards.

With deadlock detection, we could consider just using standard shard resource locks to ensure we don't update replicas in a different order, and allowing concurrent multi-shard commands.

The main downside of removing the locks is that some concurrency scenarios will throw errors where they currently just work (albeit slowly), and you're more likely to hit connection limits due to using connection per placement.

jasonmp85 commented:
Hey @marcocitus, I addressed your latest feedback and fixed some issues that arose around the rebase on top of @metdos's changes. I think everything in the code itself is addressed now. Can you give the C code a final look (if you're back in the office) and let me know how it looks? I know I still haven't gotten the tests added but I'll be able to get to that tomorrow and wondered if I could get a provisional :shipit: pending tests for various cases. Between meetings, the rebase, and a breaking change I hadn't noticed around `copyObject` I didn't get to them today.

marcocitus commented:
Yep, consider my previous messages a provisional :shipit:

anarazel commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -1994,6 +2024,129 @@ IntersectPlacementList(List *lhsPlacementList, List *rhsPlacementList)
+	ListCell *insertValuesCell = NULL;
+	List *modifyRouteList = NIL;
+
+	insertValuesList = SortList(insertValuesList, CompareInsertValuesByShardId);
```
Is that a good idea? `qsort`, which is what `SortList` uses, isn't guaranteed to be stable - in fact . Therefore sorting like this has a good chance of removing the ordering of individual rows, which in turn can cause problems because the ordering might have been intention for deadlock avoidance purposes.
Seems better to have a `shardId->List hashmap`, a `shardOffset` indexed array of lists, or a `bsearch()`able array (not list!) of shards (think I'd go with the last variant), that things get split over. As that makes the loop body simpler, I don't think you'll even end up with more code.  As an aside, it avoids the ridiculous practice of turning a list into an array, then back into a list ;)

marcocitus replied:
Ordering guarantees are lost as soon as the insert spans multiple shards. For a single shard it might be useful.

jasonmp85 replied:
Good/subtle catch. I didn't realize that users might intentionally craft the order of multi-row INSERTs to avoid unnecessary deadlocks. Adding a secondary sort key (list index) for now to add stability to the algorithm. Fixing in a better way tomorrow morning.

jasonmp85 replied:
I still think it's worthwhile that if two rows are sorted in a particular order in the initial statement that they maintain that order if sent to the same shard, so fixed it.

anarazel commented on src/backend/distributed/planner/multi_router_planner.c:
```diff
@@ -2170,3 +2323,31 @@ get_all_actual_clauses(List *restrictinfo_list)
+	int64 leftShardId = leftValue->shardId;
+	int64 rightShardId = rightValue->shardId;
+
+	/* we compare 64-bit integers, instead of casting their difference to int */
```
That comment imo is cargo culting. The previous practice was buggy, so I don't see why it makes sense to enshrine it in such comments.

jasonmp85 replied:
Removing the comment.
